### PR TITLE
added random offsets to first task intervals

### LINKED ISSUE #1159: random offset for schedule_repeated
When adding multiple repeated tasks which have the same interval, all of these will start at the same time, and it will take multiple iterations until they may distribute evenly. A single server restart may again lead to the problem (i.e., all 30-minute interval tasks will start at the same time).

This can be done better by adding a random offset to schedule-repeated.

### PR DESCRIPTION
**Pull request by t-gerhard**
Closes #1159

### CODE REVIEW

File: shared/lib/tasks.py [hunk 1]
```diff
@@ -19,7 +19,22 @@ def __init__(self, fn, args=None, kwargs=None, timeout=0, repeated=False, immedi
 		self.fn = fn
 		self.args = args
 		self.kwargs = kwargs
-		self.next = time.time() + (0 if immediate else timeout) 
+
+		# if random offset to be used as first timeout, next_timeout must be randomly selected.
+		if random_offset:
+			next_time = random.randint(0, timeout)
```
Comment (dswd) [inline shared/lib/tasks.py]: timeout and next_time are floats, so: `random.random() * timeout`

File: shared/lib/tasks.py [hunk 2]
```diff
@@ -19,7 +19,22 @@ def __init__(self, fn, args=None, kwargs=None, timeout=0, repeated=False, immedi
...
+		else:
+			next_time = timeout
+
+		# if there is an additional, immediate execution, next_time has to be used as next_timeout, and next_time must be 0.
+		if immediate:
+			self.next_timeout = next_time  # will be overwritten by self.timeout after first execute()
```
Comment (dswd) [inline shared/lib/tasks.py]: There is no self.next_timeout, this is never used

File: shared/lib/tasks.py [hunk 3]
```diff
@@ -19,7 +19,22 @@ def __init__(self, fn, args=None, kwargs=None, timeout=0, repeated=False, immedi
...
+			next_time = 0
+		else:
+			self.next_timeout = timeout
```
Comment (dswd) [inline shared/lib/tasks.py]: `next_time = timeout`

File: shared/lib/tasks.py [hunk 4]
```diff
@@ -33,7 +48,8 @@ def execute(self):
 		self.duration = time.time()-self.last
 		self.last = self.next
 		if self.repeated:
-			self.next = time.time() + self.timeout
+			self.next = time.time() + self.next_timeout
+			self.next_timeout = self.timeout
```
Comment (dswd) [inline shared/lib/tasks.py]: Why?

File: shared/lib/tasks.py (hunk 2 above)
Comment (t-gerhard) [inline shared/lib/tasks.py]: ```next_time``` determines the first execution.
```self.next_timeout``` determines the time between first and second execution.
```self.timeout``` determines the time between all other execution (2nd-3rd, 3rd-4th, ...)

next_timeout is needed. 4 cases:

* not immediate and not random_offset:
 * first execution at now+timeout, then repeat every timeout.
 * next_time == timeout and self.next_timeout == timeout
* not immediate and random_offset:
 * first execution at now + random_time, then repeat every timeout.
 * next_time == random() and self.next_timeout == timeout
* immediate and not random_offset:
 * first execution immediately. second execution a timeout later. then repeat every timeout.
 * next_time == 0 and self.next_timeout == timeout
* immediate and random_offset:
 * first execution immediately. second execution a random_time later. then repeat every timeout.
 * next_time == 0 and self.next_timeout == random()

self.next_timeout will be set to self.timeout in line 52. So this is used exactly once when rescheduling the Task after its execution in line 51.

File: shared/lib/tasks.py (hunk 4 above)
Comment (t-gerhard) [inline shared/lib/tasks.py]: See my comment above. this makes sure that self.next_timeout is only used once, and self.timeout is used for all further executions.

**Pull request merged**